### DISCUSSION TITLE
Exploring custom cost functions

### DISCUSSION THREAD

**Discussion by DaniJonesOcean** [great-lakes-ai-lab/cnbs-predictor#4] [General]
### **Overview**
As discussed previously, we would like to optimize our Gaussian Process (GP) predictive algorithm, currently implemented in `scikit-learn`, by exploring different cost functions and optimization algorithms. This project idea involves evaluating multiple metrics.

This project was inspired by a poster at AGU24 related to NOAA OWP activities: 
`H31L-0759: Exploring calibration of NextGen formulations using various objective functions and optimization algorithms`, presented by Yuqiong Liu, Earth Resources Technology Inc.

### **Objectives**
- **Explore Custom Cost Functions:** Develop and implement custom cost functions tailored to our specific needs in water supply modeling.
- **Optimize GP Hyperparameters:** Utilize different optimization algorithms (e.g., Dynamically Dimensioned Search [DDS], Grey Wolf Optimization [GWO], and Particle Swarm Optimization [PSO]) to fine-tune GP hyperparameters based on the defined cost functions.
- **Evaluate Performance Metrics:** Assess calibration and validation performance using a set of evaluation metrics.

### **Optimization Algorithms**
| **Algorithm** | **Key Features** | **Advantages** | **Best Use Cases** |
|---------------|-------------------|-----------------|---------------------|
| **DDS**       | Dynamic dimension reduction, single-solution based, stochastic component | Efficient for high-dimensional problems, minimal parameter tuning, simple implementation | High-dimensional optimization, limited computational budget, fine-tuning required |
| **GWO**       | Inspired by grey wolf social hierarchy and hunting behavior | Good exploration-exploitation balance, avoids local optima, minimal parameter tuning | Multimodal problems, diverse solution sets required |
| **PSO**       | Swarm-based movement influenced by individual and collective experiences | Fast convergence, natural parallelism, widely applicable | Low-dimensional, smooth optimization landscapes, quick convergence needed |

#### **Standard Metrics**
- **CORR:** Measures linear association between observation & simulation.
- **KGE (Kling-Gupta Efficiency):** Combines correlation, bias, & variability.
- **NSE (Nash-Sutcliffe Efficiency):** Assesses relative magnitude of residual variance compared to measured data variance.
- **NSELOG:** Emphasizes prediction of lower flows with logarithmic transformation.
- **NNSE (Normalized NSE):** \(1 / (2 - \text{NSE})\), normalizes NSE between 0 and 1.
- **RSR (RMSE-Observation Standard Deviation Ratio):** Goodness-of-fit measure similar to NSE.
- **RMSE (Root Mean Squared Error):** Average difference between prediction & observation, emphasizing larger errors.
- **MAE (Mean Absolute Error):** Similar to RMSE but without emphasizing larger errors.
- **PBIAS (Percent Bias):** Mean percent difference between prediction & observation.

#### **FDC-Based Metrics**
- **HSEG_FDC:** Percent bias of high-segment volume in flow duration curve.
- **LSEG_FDC:** Percent bias of low-segment volume in flow duration curve.

#### **Categorical Metrics**
- **POD (Probability of Detection):** Fraction of observed events correctly predicted.
- **FAR (False Alarm Ratio):** Fraction of predicted events not observed.
- **CSI (Critical Success Index):** Measures accuracy in predicting events above threshold.

#### **Event-Based Metrics**
- **PKBIAS:** Median absolute percent bias in peak flow magnitude for events above threshold.
- **PKTE:** Median absolute error in peak timing for events above threshold.
- **EVBIAS:** Median absolute percent bias in the volume of events above threshold.

### **Calibration Efficiency Observations**
Some takeaways from the AGU poster that inspired this mini-project:
- **Dependency:** Calibration efficiency during validation depends on the chosen objective function, optimization algorithm, and basin characteristics.
- **Metric Calibration:** Best validation results for a specific metric are often achieved by calibrating with a different metric (e.g., best KGE via NSE).

### **Tasks**
1. **Define Custom Cost Functions**
   - Develop cost functions that align with water supply modeling objectives.
   - Example:
     ```python
     def custom_cost_function(y_true, y_pred):
         return np.mean(np.abs(y_true - y_pred))  # Replace with specific metric
     ```
2. **Implement Optimization Algorithms**
   - **DDS:** Implement or integrate an existing DDS library.
   - **GWO:** Implement or integrate a Grey Wolf Optimizer.
   - **PSO:** Implement or integrate a Particle Swarm Optimizer.
   
3. **Integrate with Gaussian Process**
   - Modify the GP fitting process to use custom cost functions.
   - Example workflow using `scipy.optimize`:
     ```python
     from sklearn.gaussian_process import GaussianProcessRegressor
     from sklearn.gaussian_process.kernels import RBF, ConstantKernel as C
     from scipy.optimize import minimize
     
     def evaluate_gp(hyperparams, X_train, y_train, X_val, y_val):
         length_scale, noise_level = hyperparams
         kernel = C(1.0) * RBF(length_scale=length_scale)
         gp = GaussianProcessRegressor(kernel=kernel, alpha=noise_level)
         gp.fit(X_train, y_train)
         y_pred = gp.predict(X_val)
         return custom_cost_function(y_val, y_pred)
     
     bounds = [(1e-2, 10), (1e-5, 1)]
     result = minimize(evaluate_gp, x0=[1.0, 1e-3], args=(X_train, y_train, X_val, y_val), bounds=bounds)
     optimal_hyperparams = result.x
     ```
4. **Set Up Evaluation Framework**
   - Create scripts to generate heatmaps based on normalized metric values.
   - Ensure each heatmap row corresponds to a calibration run with the objective function on the y-axis and evaluation metrics on the x-axis.
    
5. **Run Calibration and Validation**
   - Perform calibration runs using each optimization algorithm.
   - Validate the performance using the defined metrics.
    
6. **Analyze Results**
   - Compare calibration efficiency across different objective functions and optimization algorithms.
   - Identify which combinations yield the best validation performance.
   
7. **Document Findings**
   - Prepare visualizations (heatmaps, time series plots).
   - Summarize key insights and observations.

### **Resources**
- **Libraries:**
  - [scikit-learn GaussianProcessRegressor](https://scikit-learn.org/stable/modules/gaussian_process.html)

Need to identify libraries for DDS, GWO, and PSO. These are apparently available in the NOAA-OWP `ngen-cal` repository, but I am not seeing them: https://github.com/NOAA-OWP/ngen-cal

### **Notes**
- **Scalability:** If necessary, consider using subsets or sparse GPs for large datasets to reduce computational costs.
- **Parallelization:** Use parallel computing capabilities of optimization algorithms to speed up evaluations (perhaps with dask).
- **Validation Strategy:** As usual, use a separate validation set for optimization to avoid overfitting.

Please add to the discussion below!
